### DISCUSSION TITLE
JSON Schema for Fungible Tokens

### DISCUSSION THREAD

**Discussion by paulatcalaxy** [hiero-ledger/hiero-improvement-proposals#403] [Ideas]
Following the HIP-10 model of defining a JSON Schema for NFTs, we@calaxy would like to define the comparable for fungible tokens. 

Premise is to link from the `memo` field on the HTS token to a JSON file that itself links to metadata like a logo, whitepaper, discord, etc. By normalizing the JSON schema we simplify for wallets , explorers, etc its display in a consistent manner

Metadata includes

- Description - human readable description of the token and corresponding project
- smallestDenom - the name of the smallest denomination of these
- Creator - the entity that created the token
- Creator DID - the decentralized identifier of the creator
- Admin - if the token has an admin key, the entity that manages that key
- Logo - the logo for the token
- Type - if the logo is included, its mime type
- Website address -the address at which additional information on the token or corresponding project
- Chat/discord address - the address at which the project's community holds discussion
- Whitepaper address - the address at which the project's whitepaper can be found

We've put together a rough first draft at https://github.com/paulatcalaxy/hedera-improvement-proposal/blob/master/HIP/hip-xxxx.md

Hoping for some review and broad stroke validation before creating a PR

paul (& cooper)

**Comment by rocketmay** [comment]
**Description Length**

No comment, 500 seems fine.

**Smallest Denom**

It should be noted in the schema that smallestDenom is the symbol which will be used by applications when displaying small denominations and is not necessarily the name. For example, for HBAR smallestDenom would be entered in the metadata as "ħ" not "Tinybar"

**Logos**

Logo and Alt Logo for the token - We should strongly recommend projects provide their logo as a .svg vector file, and recommend the logo be square.

I'm not sure what the distinction between Logo and Alt Logo is, but perhaps guidelines should be given? ie. Logo suitable for light backgrounds and Alt Logo suitable for dark? 

Alt Logo should be included in the human-readable metadata (currently missing from the list). 

**Mutable Data**

Although I understand the metadata including things such as website, whitepaper and discord address, I wonder about there being an issue where due to the immutable nature of the metadata, these fields cannot be updated if they were to change in the future. This would only be mutable if the token had an admin key, which may not be desirable as noted in the HIP.

This is a tricky question because even if the metadata is hosted somewhere mutable such as the project's website, what if that project website changes? A token without an admin key would have no recourse to update their metadata.

For a project to be required to set a token admin key purely to be able to update aesthetics seems a bit overkill.

If there was a way to change the metadata separately, with a 'metadata key', that would solve the problem fairly elegantly. However that just adds more bloat to the platform. Or perhaps the treasury could by default have permission to update the metadata and only the metadata. Not sure how feasible or reasonable that is either.

**Properties Field**

I concur with providing an optional "properties" field, much like HIP-10, for arbitrary data that a FT might want to include (for example, website/discord/whitepaper), if only to capture unknown, unanticipated future use cases and additional information a project might want to include.

**Metadata field in FT's**

I support this idea, it depends on Hedera to make the call regarding scalability of adding another field.

**Reply by bluemonad** [reply]
This looks great. It might be helpful to have an optional field for other metadata that the token creator might want to add (that aren't a textual description or a website). Also, perhaps it would be useful to allow an optional `image` property, similar to HIP-10.

If the intended meaning of `smallestDenom` is as the first reply states, then maybe a different name should be chosen? Intuitively, I would guess most people would think `TinyBar` is the `smallestDenom` of HBAR and not "ħ". Alternatively, another field could be added for the default textual representation of the denomination.

**Reply by paulatcalaxy** [reply]
> 

thanks May, I will add the generic `properties `field to follow the HIP-10 model. 

We have had preliminary discussions with Hedera on the value of a consistent mechanism across NFTS and FTs for pointing to metadata. I expect the original HTS API just didnt anticipate the value of FT metadata

**Comment by solubrew** [comment]
Why not use an IPFS type link to a place for mutable meta data?

**Reply by Cooper-Kunz** [reply]
Hey @solubrew - we're attempting to define a JSON schema that could be hosted on IPFS, Arweave, the Hedera network itself (the Hedera File Service, i.e. HFS) to enable what we see as the full spectrum of immutable <> partially mutable <> fully mutable. You can see an example of the relationships between the token, JSON storage, and what's referenced in the JSON itself within the draft proposal which I'll reshare here

> Mutability of the metadata for a token will depend on a) whether or not the token itself has an admin key (and so the link to metadata via the `'memo' can be changed to another different JSON metadata file and b) whether the storage location of the JSON metadata file guarantees immutability (e.g. IPFS of HFS), and c) whether the storage location of any media the JSON references guarantees immutability. 

> This dependence is represented below

```
-----------------            ---------------------           -------------------
|    Hedera      |          |    JSON storage    |           |   media storage  |
|                |          |                    |           |                  |
|  --------      |          |    --------        |           |                  |
|  | token |     |          |    | JSON  |       |           |                  |
|  | <memo>|-----|----------|--->|       |       |           |    -----------   |
|  |       |     |          |    | <logo>|-------|-----------|--->| logo file|  |
|   -------      |          |     -------        |           |    |__________|  |
|________________|          |____________________|           |__________________|  
   mutable?                      mutable?                         mutable?
```

This model allows flexibility to store the JSON on IPFS, and that JSON could point to a mutable source like google cloud, for example, to store the assets/images/whitepaper/etc. Or you could choose to store the entire JSON object  on something like google cloud. 

At Calaxy we likely would expect to store the JSON on IPFS as well as the asset/media itself. Other links like discord/whitepaper would likely be traditional URLs that's contents are not immutable, and thus could be updated whenever a new version of a whitepaper is released.

**Reply by solubrew** [reply]
Thanks for the detailed explanation.  This thread caught my eye because I've been thinking about some of the same items you refer too although my approach would be to create an NFT on-chain of the minimal unmutable data with links to the IPFS/AR/HFS for the mutable stuff.  I believe Fungible Token Metadata needs to be found on-chain at least a minimal amount with on-chain links to offchain resources.

**Reply by Cooper-Kunz** [reply]
That is an interesting consideration @solubrew. If I'm understanding you correctly you're suggesting something like:

Fungible token -> (points to) NFT (storing media, i.e. a logo) -> JSON (description, links, etc.) -> potentially other media/storage (?)

I'd see that as an anti-pattern - you shouldn't have to create a new token to store metadata about a different token... In our proposal, the transaction memos could refer to a file that is stored on-chain through the Hedera File Service (with optional flexibility to store off-chain entirely). In your scenario wishing for more on-chain resources, you could also store media assets (like an .svg), the description, or really most things, on the Hedera File Service as well (within the service's storage limitations). 

More specifically on the trade-offs/why I see it as an anti-pattern: it seems as though storing fungible token metadata in the form of an NFT would be unnecessary from a cost perspective (extra $1 in network fees to create the NFT for every fungible token created), doesn't need transferability that NFTs typically imply, complicates wallet storage / ownership, and provides no additional flexibility for storing the metadata on/off chain than the proposal as it stands - it attempts to allow 100% on-chain, 100% off-chain, and as much flexibility in between as possible.

**Reply by paulatcalaxy** [reply]
additionally, a wallet resolving JSON from an NFT (as in the above flow) might expect to parse HIP-10 style JSON and not the FT optimized JSON this spec defines

**Reply by solubrew** [reply]
I see your points, the one reason I think that the NFT link needs to be in the sequence is that currently building any type of application that interacts with large quantities of fungible tokens becomes an effort of piecing together lists from various githubs and other services.  Where if all FTs had a single collection contract which could have an NFT minted with base details for that FT protocol then applications could be built with a single source to access possible other protocols.  Of course this single location can be created off-chain but not in decentralized permissionless way.

**Reply by Cooper-Kunz** [reply]
> interacts with large quantities of fungible tokens becomes an effort of piecing together lists from various githubs and other services. 

I believe that the attempt at standardizing the JSON schema, embedded on-chain & attached auto the token, is addressing this problem and sufficiently reduces the complexity required versus alternatives. The "single source of truth is the tokens themselves on-chain (and their potentially off-chain (but well defined and standardly structured) metadata). 

If you (and the community) think that the standard needs to accommodate NFTs then we could potentially include an optional NFT reference in the schema - however I'd still argue that is an anti-pattern that should likely not be encouraged. e.g. 

```
{
    "description": "description - RECOMMENDED - max. of 500 characters",
    "smallestDenom" : "how to refer to the smallest denomination of these tokens - OPTIONAL",
    "creator": "creator(s) - RECOMMENDED",
    "creatorDID": "DID  - OPTIONAL ",
    "admin": "admin(s) - OPTIONAL",
    "logo": "IPFS CID or path to the token's logo file - RECOMMENDED",
    "altLogo": "IPFS CID or path to the token's alternative logo file - OPTIONAL",
    "type": "mime type - i.e. image/jpeg - CONDITIONALLY OPTIONAL "
    "website": "link to website -  OPTIONAL", 
    "discussion": "link to discussion/discord -  OPTIONAL", 
    "whitepaper": "link to whitepaper -  OPTIONAL",
    "nftRef": "HTS NFT Token ID - OPTIONAL",
}
```

**Reply by bluemonad** [reply]
This usage of an NFT as a single source of truth sounds like a rather specific choice of developers. Should this be part of the schema? A general "other" field would allow people to choose if they would like to implement something like this (or something else).

**Reply by paulatcalaxy** [reply]
Ive updated the draft to 

- add a properties container for arbitrary data
- cleaned up the smallest unit definition to distinguish between a name and a symbol
- cleaned up the definition of logo fields to make explicit the differentiation between light & dark logos

**Reply by paulatcalaxy** [reply]
https://github.com/hashgraph/hedera-improvement-proposal/pull/405

**Comment by rocketmay** [comment]
I was chatting with @sergmetelin and the topic of KYC came up - that if a FT has a KYC requirement, it's difficult to put that requirement into context without additional data. We could consider adding in a standardized optional field in properties for a KYC website/url so that tokens with KYC requirements can direct users to the appropriate information?

"Properties": {
"KYCAddress":"url to kyc information"
}

**Reply by paulatcalaxy** [reply]
@rocketmay I agree that it would be useful to have an optional KYCAddress but will propose it belongs as a peer of the other denominated fields versus a property. Thoughts?

**Reply by rocketmay** [reply]
my thought was that KYC is probably niche enough that it makes more sense to have be a property rather than store it in the root of the token, to keep the root clean. In my mind root is for fields which have broad applicability across many token types.

However, I also see the argument for keeping it in root, as it is indeed related to an intrinsic attribute of the token on the platform. 

I don't have strong feelings either way.
